apps/web/src/state/farmsV4/state/poolApr/fetcher.ts

```diff
@@ -61,20 +62,60 @@ export const getV3PoolCakeApr = async (
 
   if (!hasCache) masterChefV3CacheMap.set(pool.chainId, { totalAllocPoint, latestPeriodCakePerSecond, poolInfo })
 
-  const cakePerYear = (BigInt(SECONDS_PER_YEAR) * latestPeriodCakePerSecond) / BigInt(1e18) / BigInt(1e12)
+  const cakePerYear = new BigNumber(SECONDS_PER_YEAR)
+    .times(latestPeriodCakePerSecond.toString())
+    .dividedBy(1e18)
+    .dividedBy(1e12)
   const cakePerYearUsd = cakePrice.times(cakePerYear.toString())
   const [allocPoint, , , , , totalLiquidity, totalBoostLiquidity] = poolInfo
   const poolWeight = new BigNumber(allocPoint.toString()).dividedBy(totalAllocPoint.toString())
+  const liquidityBooster = new BigNumber(totalBoostLiquidity.toString()).dividedBy(totalLiquidity.toString())
 
-  const baseApr = cakePerYearUsd
-    .times(poolWeight)
-    .dividedBy(pool.tvlUsd ?? 1)
-    .times(totalLiquidity.toString())
-    .dividedBy(totalBoostLiquidity.toString())
+  const baseApr = cakePerYearUsd.times(poolWeight).dividedBy(liquidityBooster.times(pool.tvlUsd ?? 1))
 
   return {
     value: baseApr.toString() as `${number}`,
     boost: baseApr.times(2).toString() as `${number}`, //
+    cakePerYear,
+    poolWeight,
+  }
+}
+
+const calcV3PoolApr = ({
+  pool,
+  cakePrice,
+  totalAllocPoint,
+  latestPeriodCakePerSecond,
+  poolInfo,
+}: {
+  pool: V3PoolInfo
+  cakePrice: BigNumber
+  totalAllocPoint: bigint
+  latestPeriodCakePerSecond: bigint
+  poolInfo: readonly [bigint, `0x${string}`, `0x${string}`, `0x${string}`, number, bigint, bigint]
+}) => {
+  const cakePerYear = new BigNumber(SECONDS_PER_YEAR)
+    .times(latestPeriodCakePerSecond.toString())
+    .dividedBy(1e18)
+    .dividedBy(1e12)
+  const cakePerYearUsd = cakePrice.times(cakePerYear.toString())
+  const [allocPoint, , , , , totalLiquidity, totalBoostLiquidity] = poolInfo
```

Comment:
can poolInfo define as a map instead of array, this way may complex to maintain

Reply:
nope, it is the contract call response, there are no `Object` concept in solidity

but it has strong type with ABI defined

Reply:
got it